commit 8c185c85880765e1d30fb4e7f9830e725f412960 (refs/changes/93/1093/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-12-08T12:53:43-08:00 (2 years, 10 months ago)
    
    DOCKER-984 'docker pull some-unreachable-ip/name:tag' takes a LONG time to fail

diff --git a/lib/backends/sdc/images.js b/lib/backends/sdc/images.js
index 60cb190..31ab93b 100644
--- a/lib/backends/sdc/images.js
+++ b/lib/backends/sdc/images.js
@@ -1410,6 +1410,14 @@ function pullImage(opts, callback) {
         opts.req.log.info({err: err, imgName: imgName,
             jobUuid: job && job.uuid}, 'imageCreate job error');
 
+        /*
+         * `recognized: false` is log.info'd below to indicated `docker pull`
+         * errors that are ones that we are *not* handling specially and
+         * falling back to "Error pulling image: Internal error". I.e. the
+         * point is to get back to those and improve the user experience.
+         */
+        var recognized = true;
+
         var errmsg;
         if (err.code === 'ENOTFOUND') {
             /* BEGIN JSSTYLED */
@@ -1448,6 +1456,9 @@ function pullImage(opts, callback) {
         } else if (err.code === 'DownloadError') {
             // E.g. A `DownloadError` from docker-registry-client.
             errmsg = format('Error downloading %s: %s', imgName, err.message);
+        } else if (err.code === 'ConnectTimeoutError') {
+            errmsg = format('Timeout connecting to host %s',
+                opts.rat.index.name);
         } else if (err.code === 'NotImplemented') {
             // E.g. OAuth auth to a Docker Registry before DOCKER-771.
             errmsg = format('Could not pull from registry %s: %s',
@@ -1466,6 +1477,7 @@ function pullImage(opts, callback) {
                 /* pass */
             }
             if (!regErr || !regErr.errors) {
+                recognized = false;
                 errmsg = 'Error pulling image: Internal error';
             } else if (regErr.errors.length === 1) {
                 var code = regErr.errors[0].code;
@@ -1488,6 +1500,11 @@ function pullImage(opts, callback) {
                 errmsg = format('Error pulling image: %s', errmsgs);
             }
         }
+
+        opts.req.log.info({err: err, imgName: imgName,
+            jobUuid: job && job.uuid, recognized: recognized, errmsg: errmsg},
+            'pullImage error');
+
         /*
          * Note: docker/pkg/jsonmessage/jsonmessage.go describes an optional
          * 'errorDetail.code' field, though I don't know specific usage of
diff --git a/package.json b/package.json
index 2131f08..0bbd4fc 100644
--- a/package.json
+++ b/package.json
@@ -1,12 +1,12 @@
 {
   "name": "sdc-docker",
-  "version": "0.4.0",
+  "version": "0.4.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
     "assert-plus": "1.0.0",
     "async": "0.9.0",
-    "docker-registry-client": "3.2.4",
+    "docker-registry-client": "3.2.6",
     "bunyan": "1.8.5",
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
     "fwrule": "git+https://github.com/joyent/sdc-fwrule.git#d1174be",
